File: web/src/media/animation.ts
import { Resource } from '../resource.js';
import type { GameImage } from '../resource.js';

export class Animation {
  public readonly frameWidth: number;
  public readonly frameHeight: number;
  private readonly frames: GameImage[];

  constructor(filename: string, numFrames?: number) {
    const allFrames = Resource.getBitmap(filename);
    this.frameHeight = allFrames.height;

    if (numFrames !== undefined) {
      this.frameWidth = Math.floor(allFrames.width / numFrames);
    } else {
      this.frameWidth = this.frameHeight;
      numFrames = Math.floor(allFrames.width / this.frameWidth);
    }

    this.frames = [];
    for (let i = 0; i < numFrames; i++) {
      this.frames.push(allFrames.subImage(i * this.frameWidth, 0, this.frameWidth, this.frameHeight));
    }
  }

  getFrame(index: number): GameImage {
    return this.frames[Math.abs(index % this.frames.length)]!;
  }

  getFrameWidth(): number { return this.frameWidth; }
  getFrameHeight(): number { return this.frameHeight; }
  getGameImage(): GameImage { return this.frames[0]!; }

  drawFrame(ctx: CanvasRenderingContext2D, frame: number, x: number, y: number,
            hFlip?: boolean, vFlip?: boolean): void {
    if (x > 320 || y > 240 || x + this.frameWidth < 0 || y + this.frameHeight < 0) return;

    const needTransform = hFlip || vFlip;
    if (needTransform) ctx.save();

    if (hFlip) {
      ctx.translate(this.frameWidth + x * 2, 0);
      ctx.scale(-1, 1);
    }
    if (vFlip) {
      ctx.translate(0, this.frameHeight + y * 2);
      ctx.scale(1, -1);
    }

    const img = this.getFrame(frame);
    img.draw(ctx, x, y);

    if (needTransform) ctx.restore();
  }
}

File: wii/src/media/animation.cpp
#include "media/animation.h"
#include "constants.h"
#include <cstdlib>

Animation::Animation(const std::string& filename, int numFrames) {
    GameImage allFrames = Resource::getBitmap(filename);
    frameHeight_ = allFrames.height;

    if (numFrames > 0) {
        frameWidth_ = allFrames.width / numFrames;
    } else {
        frameWidth_ = frameHeight_;
        numFrames = (frameWidth_ > 0) ? allFrames.width / frameWidth_ : 1;
    }

    if (numFrames <= 0) numFrames = 1;

    for (int i = 0; i < numFrames; i++) {
        frames_.push_back(allFrames.subImage(i * frameWidth_, 0, frameWidth_, frameHeight_));
    }
}

GameImage Animation::getFrame(int index) const {
    if (frames_.empty()) return {};
    return frames_[std::abs(index) % static_cast<int>(frames_.size())];
}

void Animation::drawFrame(SDL_Renderer* renderer, int frame, int x, int y,
                          bool hFlip, bool vFlip) const {
    if (x > SCREEN_WIDTH || y > SCREEN_HEIGHT ||
        x + frameWidth_ < 0 || y + frameHeight_ < 0) return;

    GameImage img = getFrame(frame);
    if (!img.texture) return;

    SDL_Rect src = {img.sx, img.sy, img.sw, img.sh};
    SDL_Rect dst = {x, y, img.sw, img.sh};

    SDL_RendererFlip flip = SDL_FLIP_NONE;
    if (hFlip) flip = static_cast<SDL_RendererFlip>(flip | SDL_FLIP_HORIZONTAL);
    if (vFlip) flip = static_cast<SDL_RendererFlip>(flip | SDL_FLIP_VERTICAL);

    SDL_RenderCopyEx(renderer, img.texture, &src, &dst, 0.0, nullptr, flip);
}
